### PR TITLE
fix: add last seen eth block into the banking checkpoint so we can re…

### LINKED ISSUE #7778: Checkpoints: potential loss of deposits during downtime
# Problem encountered
https://github.com/vegaprotocol/vega/issues/7751 handles the case where any deposits made during network downtime would be lost when the network was restored from a snapshot. This could happen, for example during a protocol-upgrade.

It is _suspected_ that the same thing could happen with checkpoints since the last seen eth block for the collateral bridge is not saved in the checkpoint. The flow would be:
- a checkpoint is taken
- time moves forward and a deposit it made which is picked up by the network
- we start the network from the checkpoint
- the deposit is not seen after the restart, and is never picked up by the vega chain.

It would be good to have a system-test to confirm this to be true.

# Observed behaviour
A clear and concise description of how the system is behaving.

# Expected behaviour

# System response
Describe what the system response was, include the output from the command, automation, or else.

# Steps to reproduce

## Manual
Steps to reproduce the behaviour manually:

## Automation
Link to automation and explanation on how to run it to reproduce the problem/bug

# Evidence

## Logs
If applicable, add logs and/or screenshots to help explain your problem.

## Additional context
Add any other context about the problem here including; system version numbers, components affected.

# Definition of Done
>ℹ️ Not every issue will need every item checked, however, every item on this list should be properly considered and actioned to meet the [DoD](https://github.com/vegaprotocol/vega/blob/develop/DEFINITION_OF_DONE.md).

**Before Merging**
- [ ] Code refactored to meet SOLID and other code design principles
- [ ] Code is compilation error, warning, and hint free
- [ ] Carry out a basic happy path end-to-end check of the new code
- [ ] All APIs are documented so auto-generated documentation is created
- [ ] All bug recreation steps can be followed without presenting the original error/bug
- [ ] All Unit, Integration and BVT tests are passing
- [ ] Implementation is peer reviewed (coding standards, meeting acceptance criteria, code/design quality)
- [ ] Create [front end](https://github.com/vegaprotocol/token-frontend/issues) or [console](https://github.com/vegaprotocol/console/issues) tickets with feature labels (should be done when starting the work if dependencies known i.e. API changes)

**After Merging**
- [ ] Move development ticket to `Done` if there is **NO** requirement for new [system-tests](https://github.com/vegaprotocol/system-tests/issues)
- [ ] Resolve any issues with broken system-tests
- [ ] Create [documentation](https://github.com/vegaprotocol/documentation/issues) tickets with feature labels if functionality has changed, or is a new feature

### PR DESCRIPTION
**Pull request by wwestgarth**
closes #7778 

We now store the last-seen collateral contract block height, pending deposits, and seen deposit hashes in the checkpoints. 

I've tried to share the logic and types for the serialisation of the asset actions between the snapshots and checkpoints, but it is a bit awkward because they are similar but not identical.

System tests passing where we make a deposit, then restore from a checkpoint taken at an earlier point in time:
https://jenkins.ops.vega.xyz/blue/organizations/jenkins/common%2Fsystem-tests-wrapper/detail/system-tests-wrapper/58484/pipeline

### CODE REVIEW

File: core/banking/checkpoint.go [hunk 1]
```diff
@@ -33,7 +37,20 @@ func (e *Engine) Checkpoint() ([]byte, error) {
 		TransfersAtTime:    e.getScheduledTransfers(),
 		RecurringTransfers: e.getRecurringTransfers(),
 		BridgeState:        e.getBridgeState(),
+		LastSeenEthBlock:   e.lastSeenEthBlock,
 	}
+
+	msg.SeenRefs = make([]string, 0, e.seen.Size())
```
[the remark this answers is not in the document]
Comment (wwestgarth) [reply core/banking/checkpoint.go:43]: No, and we don't in the snapshot code path either. Do you suspect the tree-traversal would be non-deterministic?

[the remark this answers is not in the document]
Comment (ze97286) [reply core/banking/checkpoint.go:43]: idk, is it guaranteed in the spec to be deterministic? collections in go be weird...

**Review by ze97286 — APPROVED** [review]

**Review by ze97286 — COMMENTED** [review]
it would be good to test this with the mainnet lnl pipeline.

**Comment by wwestgarth** [comment]
> it would be good to test this with the mainnet lnl pipeline.

Nice and green:
https://jenkins.ops.vega.xyz/blue/organizations/jenkins/common%2Fsystem-tests-lnl-mainnet/detail/system-tests-lnl-mainnet/276/pipeline/530

**Pull request merged**